lib/sentry/envelope.ex

```diff
@@ -0,0 +1,186 @@
+defmodule Sentry.Envelope do
+  @moduledoc false
+
+  alias Sentry.{Config, Event, Util}
+
+  @type t :: %__MODULE__{
+          event_id: String.t()
+        }
+
+  defstruct event_id: nil, items: []
+
+  @doc """
+  Creates a new empty envelope.
+  """
+  @spec new() :: t()
+  def new() do
+    %__MODULE__{
+      event_id: Util.uuid4_hex(),
+      items: []
+    }
+  end
+
+  @doc """
+  Adds an event to the envelope.
+  """
+  @spec add_event(t(), Event.t()) :: t()
+  def add_event(envelope, %{event_id: event_id} = event) do
+    envelope
+    |> Map.put(:event_id, event_id)
+    |> Map.update!(:items, fn items ->
+      items ++ [event]
+    end)
+  end
+
+  @doc """
+  Encodes the envelope into it's binary representation.
+  """
+  @spec to_binary(t()) :: {:ok, String.t()} | {:error, any()}
+  def to_binary(envelope) do
+    buffer = encode_headers(envelope)
+
+    # write each item
+    Enum.reduce_while(envelope.items, {:ok, buffer}, fn item, {:ok, acc} ->
+      # encode to a temporary buffer to get the length
+      case encode_item(item) do
+        {:ok, encoded_item} ->
+          length = byte_size(encoded_item)
+          type_name = item_type_name(item)
+
+          {:cont,
+           {:ok,
+            acc <>
+              "{\"type\":\"#{type_name}\",\"length\":#{length}}\n" <>
+              encoded_item <>
+              "\n"}}
+
+        {:error, error} ->
+          {:halt, {:error, error}}
+      end
+    end)
+  end
+
+  @doc """
+  Decodes the envelope from it's binary representation.
+  """
+  @spec from_binary(String.t()) :: {:ok, t()} | {:error, :invalid_envelope}
+  def from_binary(binary) do
+    with {:ok, {raw_headers, raw_items}} <- decode_lines(binary),
+         {:ok, headers} <- decode_headers(raw_headers),
+         {:ok, items} <- decode_items(raw_items) do
+      {:ok,
+       %__MODULE__{
+         event_id: headers["event_id"] || nil,
+         items: items
+       }}
+    else
+      _e -> {:error, :invalid_envelope}
+    end
+  end
+
+  @spec from_binary!(String.t()) :: t() | no_return()
+  def from_binary!(binary) do
+    {:ok, envelope} = from_binary(binary)
+    envelope
+  end
+
+  # Returns the event in the envelope if one exists.
+  @spec event(t()) :: Event.t() | nil
+  def event(envelope) do
+    envelope.items
+    |> Enum.filter(fn item -> is_struct(item, Event) end)
```

Comment:
Ah, I had forgotten that `is_struct/2` isn't available in `1.10.x`. Could you switch this to something like `match?(%Sentry.Event{}, %{__struct__: Sentry.Event})` ?

Reply:
Replaced in 4f8603c